tapchannel/aux_traffic_shaper.go

```diff
@@ -0,0 +1,242 @@
+	err := lfn.MapOptionZ(fundingBlob, func(blob tlv.Blob) error {
+		_, err := cmsg.DecodeOpenChannel(blob)
+		return err
+	})
+	if err != nil {
+		return false, nil
+	}
+
+	// No error, so this is a custom channel, we'll want to decide.
+	return true, nil
+}
+
+// PaymentBandwidth returns the available bandwidth for a custom channel decided
+// by the given channel aux blob and HTLC blob. A return value of 0 means there
+// is no bandwidth available. To find out if a channel is a custom channel that
+// should be handled by the traffic shaper, the HandleTraffic method should be
+// called first.
+func (s *AuxTrafficShaper) PaymentBandwidth(htlcBlob,
+	commitmentBlob lfn.Option[tlv.Blob]) (lnwire.MilliSatoshi, error) {
+
+	// This method shouldn't be called if we don't have a commitment blob
+	// available (i.e., the channel is not a custom channel).
+	if commitmentBlob.IsNone() || htlcBlob.IsNone() {
+		return 0, fmt.Errorf("no commitment or TLV blob available " +
+			"for custom channel bandwidth estimation")
+	}
+
+	commitmentBytes := commitmentBlob.UnsafeFromSome()
+	htlcBytes := htlcBlob.UnsafeFromSome()
+
+	commitment, err := cmsg.DecodeCommitment(commitmentBytes)
+	if err != nil {
+		return 0, fmt.Errorf("error decoding commitment blob: %w", err)
+	}
+
+	htlc, err := rfqmsg.DecodeHtlc(htlcBytes)
+	if err != nil {
+		return 0, fmt.Errorf("error decoding HTLC blob: %w", err)
+	}
+
+	localBalance := cmsg.OutputSum(commitment.LocalOutputs())
+
+	// There either already is an amount set in the HTLC (which would
+	// indicate it to be a keysend payment), in which case we don't need an
+	// RFQ ID as we can just compare the local balance and the required HTLC
+	// amount. If there is no amount set, we need to look up the RFQ ID in
+	// the HTLC blob and use the accepted quote to determine the amount.
+	htlcAssetAmount := htlc.Amounts.Val.Sum()
+	if htlcAssetAmount != 0 && htlcAssetAmount <= localBalance {
+		// We signal "infinite" bandwidth by returning a very high
+		// value (number of Satoshis ever in existence).
+		return lnwire.NewMSatFromSatoshis(btcutil.MaxSatoshi), nil
+	}
+
+	// If the HTLC doesn't have an RFQ ID, we can't use any channel.
+	if htlc.RfqID.ValOpt().IsNone() {
+		return 0, nil
+	}
+
+	rfqID := htlc.RfqID.ValOpt().UnsafeFromSome()
+	acceptedQuotes := s.cfg.RfqManager.PeerAcceptedSellQuotes()
+	quote, ok := acceptedQuotes[rfqID.Scid()]
+	if !ok {
+		return 0, fmt.Errorf("no accepted quote found for RFQ ID "+
+			"%x (SCID %d)", rfqID[:], rfqID.Scid())
+	}
+
+	mSatPerAssetUnit := quote.BidPrice
+
+	// The available balance is the local asset unit expressed in
+	// milli-satoshis.
+	return lnwire.MilliSatoshi(localBalance) * mSatPerAssetUnit, nil
+}
+
+// ProduceHtlcExtraData is a function that, based on the previous extra data
+// blob of an HTLC, may produce a different blob or modify the amount of bitcoin
+// this htlc should carry.
+func (s *AuxTrafficShaper) ProduceHtlcExtraData(totalAmount lnwire.MilliSatoshi,
```

Comment:
I think the method name and doc here might be outdated. This method seems only concerned with "custom records". The `ExtraData` field is used in wire message serialization and is unrelated.